src/renderer/api/kube-watch-api.ts

```diff
@@ -43,50 +42,59 @@ export interface IKubeWatchLog {
 
 @autobind()
 export class KubeWatchApi {
-  private cluster: Cluster;
-  private namespaceStore: NamespaceStore;
-
   private requestId = 0;
-  private isConnected = false;
   private reader: ReadableStreamReader<string>;
-  private subscribers = observable.map<KubeApi, number>();
-
-  // events
   public onMessage = new EventEmitter<[IKubeWatchMessage]>();
 
+  @observable.ref private getCluster: () => Cluster;
+  @observable.ref private getNamespaces: () => string[];
+  @observable isConnected = false;
+  @observable subscribers = observable.map<KubeApi, number>();
+
+  @computed get isReady(): boolean {
+    return !!(this.cluster && this.namespaces);
+  }
+
+  @computed get cluster(): Cluster | undefined {
+    return this.getCluster?.();
+  }
+
+  @computed get namespaces(): string[] {
+    return this.getNamespaces?.() || [];
+  }
+
   @computed get isActive(): boolean {
     return this.apis.length > 0;
   }
 
   @computed get apis(): string[] {
-    const { cluster, namespaceStore } = this;
-    const activeApis = Array.from(this.subscribers.keys());
-
-    return activeApis.map(api => {
-      if (!cluster.isAllowedResource(api.kind)) {
+    return Array.from(this.subscribers.keys()).map(api => {
```

Comment:
I recently learned about this, which I think is probably faster since it doesn't have to create the intermediate array.
```suggestion
    return Array.from(this.subscribers.keys(), api => {
```

Reply:
I believe that such micro-optimizations we should avoid in advance. In most cases you won't notice any performance difference..
If it brings more readability or reasonable performance boost, only then apply (but it's in that case + you must know what does the second argument => more cognitive load).